server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -0,0 +1,284 @@
+ */
+
+package org.opensearch.node.remotestore;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterName;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.blobstore.BlobMetadata;
+import org.opensearch.common.blobstore.BlobPath;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.concurrent.AbstractAsyncTask;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.gateway.remote.model.RemotePinnedTimestamps;
+import org.opensearch.gateway.remote.model.RemoteStorePinnedTimestampsBlobStore;
+import org.opensearch.index.translog.transfer.BlobStoreTransferService;
+import org.opensearch.node.Node;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.isRemoteStoreClusterStateEnabled;
+
+/**
+ * Service for managing pinned timestamps in a remote store.
+ * This service handles pinning and unpinning of timestamps, as well as periodic updates of the pinned timestamps set.
+ *
+ * @opensearch.internal
+ */
+public class RemoteStorePinnedTimestampService implements Closeable {
+    private static final Logger logger = LogManager.getLogger(RemoteStorePinnedTimestampService.class);
+    private static Tuple<Long, Set<Long>> pinnedTimestampsSet = new Tuple<>(-1L, Set.of());
+    public static final int PINNED_TIMESTAMP_FILES_TO_KEEP = 5;
+
+    private final Supplier<RepositoriesService> repositoriesService;
+    private final Settings settings;
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+    private BlobStoreRepository blobStoreRepository;
+    private BlobStoreTransferService blobStoreTransferService;
+    private RemoteStorePinnedTimestampsBlobStore pinnedTimestampsBlobStore;
+    private AsyncUpdatePinnedTimestampTask asyncUpdatePinnedTimestampTask;
+    private volatile TimeValue pinnedTimestampsSchedulerInterval;
+
+    /**
+     * Controls pinned timestamp scheduler interval
+     */
+    public static final Setting<TimeValue> CLUSTER_REMOTE_STORE_PINNED_TIMESTAMP_SCHEDULER_INTERVAL = Setting.timeSetting(
+        "cluster.remote_store.pinned_timestamps.scheduler_interval",
+        TimeValue.timeValueMinutes(3),
+        TimeValue.timeValueMinutes(1),
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
```

Comment:
I'm concerned about the settings being dynamic, it's quite possible that if this value were to be set to 1 min and the node is unable to apply the state within a specified time, it might take lag detector atleast 90s to kick the node out?

Reply:
pinned timestamps will not be part of cluster state so node will not be kicked out if it is not able to fetch pinned timestamps within X mins. Remote store GCs will be skipped if the last fetch timestamp is > X mins.

But we can keep the setting static and later think of making it dynamic.